framework/src/modules/chain/bft/bft.js

```diff
@@ -334,13 +325,64 @@ class BFT extends EventEmitter {
 	}
 
 	get maxHeightPrevoted() {
-		return this.finalityManager.prevotedConfirmedHeight;
+		return this.finalityManager.chainMaxHeightFinalized;
 	}
 
 	reset() {
 		this.finalityManager.headers.empty();
 		this.finalityManager.recompute();
 	}
+
+	async _fillCache(minActiveHeightsOfDelegates) {
+		const tillHeight = this.finalityManager.minHeight - 1;
+		const fromHeight =
+			this.finalityManager.maxHeight -
+			// Search is inclusive, therefore, it should start from one above (ex: 288 - 500, which results total 303)
+			this.constants.activeDelegates * BFT_ROUND_THRESHOLD +
+			1;
+		const blocksJSON = await this.blockEntity.get(
+			{ height_gte: fromHeight, height_lte: tillHeight },
+			{ limit: null, sort: 'height:desc' },
+		);
+		for (const blockJSON of blocksJSON) {
+			if (blockJSON.height === 1) {
+				this.finalityManager.headers.add(
+					extractBFTBlockHeaderFromBlock({
+						...blockJSON,
+						delegateMinHeightActive: 1,
+					}),
+				);
+				return;
+			}
+
+			const activeHeights =
+				minActiveHeightsOfDelegates[blockJSON.generatorPublicKey];
+			if (!activeHeights) {
+				throw new Error(
+					`Minimum active heights were not found for delegate "${blockJSON.generatorPublicKey}".`,
+				);
+			}
+
+			// If there is no minHeightActive until this point,
+			// we can set the value to 0
+			const minimumPossibleActiveHeight = this.slots.calcRoundStartHeight(
+				this.slots.calcRound(
+					Math.max(blockJSON.height - this.constants.activeDelegates * 3, 1),
+				),
+			);
+			const [delegateMinHeightActive] = activeHeights.filter(
+				height => height >= minimumPossibleActiveHeight,
+			);
+
+			const blockHeader = {
+				...blockJSON,
+				delegateMinHeightActive,
+			};
+			this.finalityManager.headers.add(
+				extractBFTBlockHeaderFromBlock(blockHeader),
+			);
```

Comment:
can we abstract this in a function and reuse in `_loadBlocksFromStorage` as well?

Reply:
it's slightly different functionality, and for the patch, i don't want to clean up and change so much.
On my other PR, it will be cleaned up